publisher/release/marker_aws_test.go

```diff
@@ -0,0 +1,643 @@
+	]`
+
+	reader := bytes.NewReader([]byte(existingMarkers))
+	readCloser := io.NopCloser(reader)
+
+	s3ClientMock.ShouldGetObject(
+		&s3.GetObjectInput{Bucket: &s3Config.Bucket, Key: aws.String(fmt.Sprintf("%s/%s", s3Config.Directory, markerName))},
+		&s3.GetObjectOutput{Body: readCloser})
+
+	startTime := time.Date(2023, 1, 2, 00, 00, 00, 0, time.UTC)
+	mark := Mark{
+		ReleaseInfo: ReleaseInfo{
+			AppName:   "another-app",
+			Tag:       "v1.2",
+			RunID:     "run3",
+			RepoName:  "repo3",
+			Schema:    "schema3",
+			SchemaURL: "url3",
+		},
+		Start: CustomTime{startTime},
+	}
+
+	markerS3 := &markerAWS{
+		client:       s3ClientMock,
+		conf:         s3Config,
+		timeProvider: timeProviderMock,
+		logfn:        nolog,
+	}
+	err := markerS3.End(mark)
+	assert.ErrorIs(t, err, ErrNoStartedMarkerFoundForApp)
+	mock.AssertExpectationsForObjects(t, s3ClientMock, timeProviderMock)
+}
+
+func Test_EndFailsIfMarkerStartTimeIsDifferent(t *testing.T) {
+	s3ClientMock := &S3ClientMock{}
+	timeProviderMock := &TimeProviderMock{}
+
+	s3Config := S3Config{
+		Bucket:    "bucket",
+		RoleARN:   "role",
+		Region:    "region",
+		Directory: "directory",
+	}
+
+	// It should read existing markers
+	existingMarkers := `
+	[
+		{"app_name": "app1", "tag": "v1.0", "run_id": "run1", "start": "2023-01-01T00:00:00Z", "end": "2023-01-01T01:00:00Z", "repo_name": "repo1", "schema": "schema1", "schema_url": "url1"},
+		{"app_name": "my-app", "tag": "v1.2", "run_id": "run3", "start": "2023-01-02T11:00:00Z", "end": "0001-01-01T00:00:00Z", "repo_name": "repo3", "schema": "schema3", "schema_url": "url3"}
+	]`
+
+	reader := bytes.NewReader([]byte(existingMarkers))
+	readCloser := io.NopCloser(reader)
+
+	s3ClientMock.ShouldGetObject(
+		&s3.GetObjectInput{Bucket: &s3Config.Bucket, Key: aws.String(fmt.Sprintf("%s/%s", s3Config.Directory, markerName))},
+		&s3.GetObjectOutput{Body: readCloser})
+
+	startTime := time.Date(2023, 1, 2, 00, 00, 00, 0, time.UTC)
+	mark := Mark{
+		ReleaseInfo: ReleaseInfo{
+			AppName:   "my-app",
+			Tag:       "v1.2",
+			RunID:     "run3",
+			RepoName:  "repo3",
+			Schema:    "schema3",
+			SchemaURL: "url3",
+		},
+		Start: CustomTime{startTime},
+	}
+
+	markerS3 := &markerAWS{
+		client:       s3ClientMock,
+		conf:         s3Config,
+		timeProvider: timeProviderMock,
+		logfn:        nolog,
+	}
+	err := markerS3.End(mark)
+	assert.ErrorIs(t, err, ErrNoStartedMarkerFoundForApp)
+	mock.AssertExpectationsForObjects(t, s3ClientMock, timeProviderMock)
+}
+
+func Test_EndFailsIfMarkerStartIsZero(t *testing.T) {
+	s3ClientMock := &S3ClientMock{}
+	timeProviderMock := &TimeProviderMock{}
+
+	s3Config := S3Config{
+		Bucket:    "bucket",
+		RoleARN:   "role",
+		Region:    "region",
+		Directory: "directory",
+	}
+
+	startTime := time.Time{}
+	mark := Mark{
+		ReleaseInfo: ReleaseInfo{
+			AppName:   "my-app",
+			Tag:       "v1.2",
+			RunID:     "run3",
+			RepoName:  "repo3",
+			Schema:    "schema3",
+			SchemaURL: "url3",
+		},
```

Comment:
(future) I see a lot of repetitions between tests. Maybe we could benefit from having some "basis" variables to reuse on the tests. Maybe is a terrible idea.

Reply:
It is a good idea :) . We could use stubs with https://github.com/brianvoe/gofakeit for example